Updated Speck implementation

clicketyclackety opened the pull request:
# Description

Currently there are 2 speck implementation, 1 local and 1 server-side. This was okay for the hackathon, but the current implementation has quite the performance penalties.

I've implemented an `ISpeck` interface to allow for more flexibility in server/client side handling of them.
In the local implementation, it de-serializes and caches the geometry to prevent constant, slow de-serialization.
In the server implementation it has all of the information it had before, but with a few different constructors.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

rfeathers068 commented on SpeckLib/ServerSpeck.cs:
```diff
@@ -0,0 +1,51 @@
+        public static ServerSpeck From(ISpeck speck)
+        {
+            return new ServerSpeck()
+            {
```
The previous implementation set Temporary = true and LockedBy = owner.  Otherwise when its passed to the database for add/update operations it will not be locked or owned.

clicketyclackety replied:
I'll push through a fix for this now

rfeathers068 commented:
> @cwensley, @rfeathers068 I can't quite work out why this doesn't work. And I don't think I'm debugging it correctly. Everything builds, but the specks aren't passed around correctly. Would love a hand with this one, either fixing, or suggesting a better implementation. 😋

One thing I noticed, which Curtis may have to help with is the Database Migration didn't update for me so that might be part of the issue as well.

![image](https://user-images.githubusercontent.com/89032297/201433517-262efd22-66be-4335-8260-33eac82d2960.png)

clicketyclackety commented:
> One thing I noticed, which Curtis may have to help with is the Database Migration didn't update for me so that might be part of the issue as well.
> 
> ![image](https://user-images.githubusercontent.com/89032297/201433517-262efd22-66be-4335-8260-33eac82d2960.png)

I had a feeling this might play a part

cwensley commented:
Hey @clicketyclackety, a few things that I can see about this PR:

1. Why is `ServerSpeck` in `SpeckLib` but `LocalSpeck` is in `Crash`?  I would expect ServerSpeck to be in Crash.Server unless there's a reason for that?  Ideally we should just use the same class on both sides and if we need other information on either side we can wrap it in an outer class.
2. I'm not sure you can use an Interface to pass things around with SignalR.. what class would it create on the server if it knows nothing about LocalSpeck?  It does not have any knowledge to create a ServerSpeck or LocalSpeck.
3. Any database model changes need to be reflected with a `dotnet ef add migration <name_of_migration>` in the `Client.Server\` folder.  I should add a VS Code task for that and some instructions to the docs.
4.  The ServerSpeck class requires a public default constructor for it to be used by the Entity Framework database, otherwise there's no way for the database calls to create instances of the class.

That's all I can see for now, I hope this helps!

clicketyclackety commented:
Hey @cwensley,

Thanks for the notes. I was able to get things running again with a few changes;
1. ServerSpeck gets used by Crash to convert into LocalSpeck. LocalSpeck uses GeometryBase and I don't want the server to have to use Rhino dlls. If you can suggest a better way to separate things, I'd love that.
2. That is really handy to know. Makes total sense. I've moved to inheritance
3. Handy command. Would love a script for it. For now I've just been cleaning the directories and I updated the initial migration
4. I've ensured default constructors and getters/setters. Everything is currently public.

clicketyclackety commented on Crash.ClientTestApp/Program.cs:
```diff
@@ -21,7 +22,7 @@ static void Main(string[] args)
             Console.WriteLine("\tc = Create Speck");
             Console.WriteLine("\tq = Quit");
 
-            var client = new CrashClient(userId, new Uri("http://localhost:5141/Crash"));
```
I found 5141 had some issues in cafes I worked in. Unsure if there is any issue with changing this as 8080 worked better. I have no strong opinions either way

clicketyclackety commented:
I think likely you guys may spot some more things. But it does now work nicely! Much nicer to be able to connect to the server and retrieve your work in progress stuff 😊

cwensley commented on Crash/LocalSpeck.cs:
```diff
@@ -0,0 +1,64 @@
+    /// <summary>
+    /// Local instance of a received speck.
+    /// </summary>
+    public sealed class LocalSpeck : Speck
```
Instead of inheriting from Speck and having to copy all of its properties, how about something like:
```
public sealed class LocalSpeck
{
  public Speck Speck { get; set; }
  public GeometryBase Geometry { get; set; }
}
```

clicketyclackety replied:
This seems like a much better idea really

clicketyclackety replied:
I did something a little different. I've created the ISpeck interface like before, and implemented it to point at the speck inside it

cwensley commented on Crash/Events/CrashInit.cs:
```diff
@@ -18,14 +19,44 @@ internal static class CrashInit
+            }
+            else
+            {
+                if (speck.LockedBy == User.CurrentUser.name)
```
I don't know if this would ever be called?  All locked specks _should_ be non-temporary..

cwensley reviewed (approved):
Looks great!

clicketyclackety commented:
@cwensley if you could take a look at the new implementation, that'd be fab 😊
